Python/traceback.c

```diff
@@ -517,9 +554,52 @@ tb_displayline(PyObject *f, PyObject *filename, int lineno, PyObject *name)
     Py_DECREF(line);
     if (err != 0)
         return err;
+    int truncation = _TRACEBACK_SOURCE_LINE_INDENT;
+    PyObject* source_line = NULL;
     /* ignore errors since we can't report them, can we? */
-    if (_Py_DisplaySourceLine(f, filename, lineno, 4))
+    if (!_Py_DisplaySourceLine(f, filename, lineno, _TRACEBACK_SOURCE_LINE_INDENT, &truncation, &source_line)) {
+        int code_offset = tb->tb_lasti;
+        if (PyCode_Addr2Line(_PyFrame_GetCode(frame), code_offset) != _PyCode_Addr2EndLine(_PyFrame_GetCode(frame), code_offset)) {
+            goto done;
+        }
+
+        int start_offset = _PyCode_Addr2Offset(_PyFrame_GetCode(frame), code_offset);
+        int end_offset = _PyCode_Addr2EndOffset(_PyFrame_GetCode(frame), code_offset);
+
+        start_offset = byte_to_character_offset_in_line(source_line, start_offset);
+        // Not sure why a `+ 1` is needed here for the end_offset.
+        end_offset = byte_to_character_offset_in_line(source_line, end_offset) + 1;
```

Comment:
@ammaraskar I found the problem. In interval notation, the offset interval is
```
[start_offset, end_offset)
```

which is closed on the right, open on the left. As we use 1-based offsets it means that the comparison for start_offset is correct (because is inclusive) but the comparison for end_offset is not because is open.  It should be:

```
        end_offset = byte_to_character_offset_in_line(source_line, end_offset - 1);
```

Can you add some tests for this?

Reply:
Yup that snippet you posted of:

```c
end_offset = byte_to_character_offset_in_line(source_line, end_offset - 1) + 1;
```

is exactly what I was thinking of. I've implemented a similar fix using `-1` to avoid the 1-indexing problem. I added some tests: https://github.com/colnotab/cpython/blob/e29545ea95c134d071a052016b010c3fd4cd83ac/Lib/test/test_traceback.py#L276-L361

But how would test this specificically/do you have suggestions for any tests for this?

Reply:
This case was failing before:

```
def foo(a: NotExistentName) -> Blech:
   ...
```

there was an off-1 error there in the highlighting. So probably we just need to add this as a test case

Reply:
Gotcha, I did add that one so we should be good then :) https://github.com/colnotab/cpython/blob/e29545ea95c134d071a052016b010c3fd4cd83ac/Lib/test/test_traceback.py#L325-L339